File: Sources/GatewayCore/ScriptDoc.swift
import Foundation

public struct Step: Equatable, Sendable {
    public enum Kind: String, Sendable {
        case say, pause, hold, media, bed, beat, level, surf, pan, use
    }
    public var kind: Kind
    public var text: String = ""
    public var seconds: Double = 0
    public var args: [Double] = []
    /// For `use`: the mode of the referenced segment, empty for the whole segment.
    public var option: String = ""

    /// Public because a step is also something callers *build*, not only
    /// something the parser hands back: the bed's cues are replayed as steps.
    public init(kind: Kind, text: String = "", seconds: Double = 0,
                args: [Double] = [], option: String = "") {
        self.kind = kind; self.text = text; self.seconds = seconds
        self.args = args; self.option = option
    }
}

extension ScriptDoc {
    /// Read a `.gws` off disk and parse it, or nil.
    ///
    /// Sixteen call sites spelled this out by hand as
    /// `(try? String(contentsOf:encoding:)).flatMap { try? ScriptParser.parse($0) }`,
    /// which is long enough that some of them dropped the `try?` handling or
    /// re-read the same file twice in one view body. Nil means "not usable" and
    /// callers already treat it that way; anywhere the *reason* matters, call
    /// `ScriptParser.parse` directly and keep the error.
    public static func load(_ url: URL) -> ScriptDoc? {
        guard let src = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        return try? ScriptParser.parse(src)
    }

    /// The source text and its parse together, for the places that need both --
    /// the editors, which show the raw text and validate it at the same time.
    public static func loadSource(_ url: URL) -> (source: String, doc: ScriptDoc?)? {
        guard let src = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        return (src, try? ScriptParser.parse(src))
    }
}

public struct ScriptDoc: Sendable {
    public var title = "untitled"
    public var level = "F10"
    public var voice = "default"
    public var ending = "return"          // return | stay
    public var seed: UInt64?
    public var pan: Double = 0
    public var beatOverride: Double?
    public var carrierOverride: Double?
    /// Segment id when this file is a segment rather than a whole session.
    public var segment: String?
    /// Structural density, 1...3. On a segment file: the level this body is
    /// authored at (1 = anchors and counts only, 2 = adds preamble and lore,
    /// 3 = full detail). On a session or template: the level requested at
    /// assembly. `nil` on a segment means the file serves every verbosity.
    /// Distinct from variants, which are the same structure phrased differently.
    public var verbosity: Int?
    /// Levels a segment is offered at. Sessions carry a single `level` instead.
    public var levels: [String] = []
    /// A placeholder body: written to be voiced and compiled now, but standing
    /// in for something only experience can supply. It must never be mistaken
    /// for a described level -- the worklist keeps counting it.
    public var provisional = false
    /// Segments in the same family are interchangeable: three forms of the
    /// Affirmation, say. Assembly picks one; none is more correct than another.
    /// Distinct from verbosity (density) and variants (phrasing).
    public var family: String?
    /// For a transition segment: the level it departs. With `@levels` as the
    /// destination this makes the segment a rung of the ladder -- relax-10
    /// declares `@from F1`, because the ten-point system IS the climb into
    /// Focus 10. Climb ids encode the same thing; `@from` states it as data.
    public var from: String?
    /// Rough spoken length, e.g. "~6m". Documentation for the person building a
    /// session; nothing computes from it.
    public var duration = ""
    /// Terms that must survive verbatim -- named Gateway instruments are
    /// terminology, not phrasing. Checked after generation, not merely requested.
    public var protectedTerms: [String] = []
    /// When true the body carries no variant groups: liturgy, not phrasing.
    public var fixed = false
    /// The authored waking exit Continuous mode offers after a listener has
    /// chosen to stay at a destination. Exactly one segment should own this
    /// role; ordinary returning sessions may end through different segments.
    public var continuousExit = false
    /// The exit to fall back on when no exit is authored for the arrival level.
    public var continuousExitDefault = false
    /// Intentionally retained source material with no active session. This is
    /// different from an authoring omission and carries the reason on disk.
    public var shelved: String?
    /// Done sitting up, before the induction — see the `@upright` directive.
    public var upright = false
    /// Things the listener needs to hand before starting, e.g. "paper", "a pen".
    /// Surfaced before a session begins, never mid-induction: being told you
    /// need a pen once you are already settled is the worst possible time.
    public var needs: [String] = []

    /// Tokens still present in spoken text. Must be empty before rendering:
    /// a line still carrying `[[destination]]` would be read out loud.
    public var unfilledTokens: [String] {
        var found: [String] = []
        for step in steps where step.kind == .say {
            var rest = Substring(step.text)
            while let open = rest.range(of: "[["),
                  let close = rest[open.upperBound...].range(of: "]]") {
                found.append(String(rest[open.upperBound ..< close.lowerBound]))
                rest = rest[close.upperBound...]
            }
        }
        return found
    }

    /// Substitute `[[key]]` tokens in every spoken line.
    ///
    /// Square brackets, deliberately: `{a|b|c}` variant groups already own
    /// braces, and a token nested inside a group would confuse the innermost-
    /// group scan. The two mechanisms never meet.
    ///
    /// This is how a segment can name something the author could not know —
    /// the verbosity the listener picked, the level they are heading for —
    /// **without the engine hardcoding anything spoken**. The frame stays in
    /// the `.gws` file where it can be read, diffed and rewritten; only the
    /// values come from data.
    ///
    /// A line still carrying a token after filling is a bug that would be
    /// *spoken aloud*, so `unfilledTokens` exists to catch it before render.
    public func filled(_ values: [String: String]) -> ScriptDoc {
        var out = self
        for i in out.steps.indices where out.steps[i].kind == .say {
            out.steps[i].text = ScriptParser.fill(out.steps[i].text, values)
        }
        return out
    }


    public var steps: [Step] = []
}

public enum ScriptError: Error, CustomStringConvertible, Equatable {
    case directiveAfterBody(String)
    case unknownDirective(String)
    case unknownVerb(String, line: String)
    case badEnding(String)
    case badNumber(String)
    case badVerbosity(String)
    case variantsInFixed(String)

    public var description: String {
        switch self {
        case .directiveAfterBody(let l): return "directive after the body started: \(l)"
        case .unknownDirective(let k):   return "unknown directive @\(k)"
        case .unknownVerb(let v, let l): return "unknown verb '\(v)' in line: \(l)"
        case .badEnding(let v):          return "@ending must be 'return' or 'stay', got '\(v)'"
        case .badNumber(let s):          return "expected a number, got '\(s)'"
        case .badVerbosity(let s):       return "@verbosity must be 1, 2 or 3, got '\(s)'"
        case .variantsInFixed(let s):    return "@fixed script contains a variant group: \(s)"
        }
    }
}

public enum ScriptParser {
    private static let panWords: [String: Double] =
        ["left": -0.9, "right": 0.9, "centre": 0, "center": 0]

    public static func parsePan(_ v: String) throws -> Double {
        let s = v.trimmingCharacters(in: .whitespaces).lowercased()
        if let w = panWords[s] { return w }
        guard let d = Double(s) else { throw ScriptError.badNumber(s) }
        return d
    }

    private static func num(_ s: String) throws -> Double {
        guard let d = Double(s.trimmingCharacters(in: .whitespaces)) else {
            throw ScriptError.badNumber(s)
        }
        return d
    }

    public static func parse(_ source: String, seedOverride: UInt64? = nil) throws -> ScriptDoc {
        var doc = ScriptDoc()
        var headerDone = false

        for raw in source.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") { continue }

            if line.hasPrefix("@") {
                if headerDone { throw ScriptError.directiveAfterBody(line) }
                let body = String(line.dropFirst())
                let parts = body.split(separator: " ", maxSplits: 1).map(String.init)
                let key = parts[0].lowercased()
                let val = parts.count > 1 ? parts[1].trimmingCharacters(in: .whitespaces) : ""
                switch key {
                case "title":    doc.title = val
                case "level":    doc.level = val
                case "voice":    doc.voice = val
                case "ending":
                    guard val == "return" || val == "stay" else { throw ScriptError.badEnding(val) }
                    doc.ending = val
                case "seed":     doc.seed = UInt64(val)
                case "beat":     doc.beatOverride = try num(val)
                case "carrier":  doc.carrierOverride = try num(val)
                case "pan":      doc.pan = try parsePan(val)
                case "segment":  doc.segment = val
                case "verbosity":
                    guard let v = Int(val), (1...3).contains(v)
                    else { throw ScriptError.badVerbosity(val) }
                    doc.verbosity = v
                case "levels":   doc.levels = val.split(separator: ",").map {
                                     $0.trimmingCharacters(in: .whitespaces) }
                case "from":     doc.from = val.uppercased()
                case "family":   doc.family = val
                case "provisional": doc.provisional = true
                case "duration": doc.duration = val
                // A task done sitting up, before the induction. Everything else
                // in the library assumes the listener is already lying down --
                // this is the one kind of step that has to happen first, and
                // may need something to hand.
                case "upright": doc.upright = true
                case "needs": doc.needs = val.split(separator: ",").map {
                    $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
                case "protected": doc.protectedTerms = val.split(separator: ",").map {
                                     $0.trimmingCharacters(in: .whitespaces) }
                case "fixed":    doc.fixed = true
                // `@continuous-exit` marks a segment as an authored waking
                // exit; its `@levels` say which arrival it was written for.
                // `@continuous-exit default` additionally makes it the one to
                // use when the arrival level has no exit of its own — the
                // ten-count is the general Gateway return, and saying so in
                // data keeps the Focus keys out of the code that chooses.
                case "continuous-exit":
                    doc.continuousExit = true
                    switch val.lowercased() {
                    case "": break
                    case "default": doc.continuousExitDefault = true
                    default: throw ScriptError.unknownDirective("continuous-exit \(val)")
                    }
                case "shelved": doc.shelved = val.isEmpty ? "intentionally inactive" : val
                default: throw ScriptError.unknownDirective(key)
                }
                continue
            }

            headerDone = true
            let parts = line.split(separator: " ", maxSplits: 1).map(String.init)
            let verb = parts[0].lowercased()
            let rest = parts.count > 1 ? parts[1].trimmingCharacters(in: .whitespaces) : ""

            switch verb {
            case "say":   doc.steps.append(Step(kind: .say, text: rest))
            case "pause": doc.steps.append(Step(kind: .pause, seconds: try num(rest)))
            case "hold":  doc.steps.append(Step(kind: .hold, seconds: try num(rest)))
            case "media":
                let f = rest.split(separator: " ").map(String.init)
                guard f.count == 2 else { throw ScriptError.badNumber(rest) }
                doc.steps.append(Step(kind: .media, text: f[0], seconds: try num(f[1])))
            case "level": doc.steps.append(Step(kind: .level, text: rest))
            case "beat":  doc.steps.append(Step(kind: .beat, args: [try num(rest)]))
            case "surf":  doc.steps.append(Step(kind: .surf, args: [try num(rest)]))
            case "pan":   doc.steps.append(Step(kind: .pan, args: [try parsePan(rest)]))
            // A template step: pull a segment in by id, optionally in a mode.
            // The template carries the order and the session-level automation;
            // the segments carry the words.
            case "use":
                let f = rest.split(separator: " ", maxSplits: 1).map(String.init)
                doc.steps.append(Step(kind: .use, text: f[0],
                                      option: f.count > 1 ? f[1].trimmingCharacters(in: .whitespaces) : ""))
            case "bed":
                let f = rest.split(separator: " ").map(String.init)
                doc.steps.append(Step(kind: .bed, args: [try num(f[0]), try num(f[1])]))
            default: throw ScriptError.unknownVerb(verb, line: line)
            }
        }

        let seed = seedOverride ?? doc.seed
        var rng = SplitMix64(seed: seed ?? 0)
        for i in doc.steps.indices where doc.steps[i].kind == .say {
            let t = doc.steps[i].text
            if doc.fixed, t.contains("{") { throw ScriptError.variantsInFixed(t) }
            doc.steps[i].text = doc.fixed ? tidy(t) : resolveVariants(t, &rng)
        }
        doc.seed = seed
        return doc
    }

    /// Collapse {a|b|c} groups, innermost first.
    static func resolveVariants(_ text: String, _ rng: inout SplitMix64) -> String {
        var out = text
        for _ in 0..<6 {
            guard let open = lastIndexOfOpen(out), let close = out[open...].firstIndex(of: "}")
            else { break }
            let inner = String(out[out.index(after: open)..<close])
            let opts = inner.split(separator: "|", omittingEmptySubsequences: false)
                            .map { $0.trimmingCharacters(in: .whitespaces) }
            let pick = opts.isEmpty ? "" : opts[Int(rng.next() % UInt64(opts.count))]
            out.replaceSubrange(open...close, with: pick)
        }
        return tidy(out)
    }

    static func fill(_ text: String, _ values: [String: String]) -> String {
        var out = text
        for (k, v) in values {
            out = out.replacingOccurrences(of: "[[\(k)]]", with: v)
        }
        return tidy(out)
    }

    /// Innermost group = the last '{' that has no '{' after it before its '}'.
    private static func lastIndexOfOpen(_ s: String) -> String.Index? {
        var found: String.Index?
        var i = s.startIndex
        while i < s.endIndex {
            if s[i] == "{" { found = i }
            if s[i] == "}" , found != nil { return found }
            i = s.index(after: i)
        }
        return nil
    }

    static func tidy(_ s: String) -> String {
        var out = s
        while out.contains("  ") { out = out.replacingOccurrences(of: "  ", with: " ") }
        return out.trimmingCharacters(in: .whitespaces)
    }

    /// Terms that must appear verbatim. Returns the ones that went missing.
    public static func missingProtectedTerms(_ doc: ScriptDoc) -> [String] {
        guard !doc.protectedTerms.isEmpty else { return [] }
        let body = doc.steps.filter { $0.kind == .say }.map(\.text).joined(separator: " ")
        return doc.protectedTerms.filter { !body.localizedCaseInsensitiveContains($0) }
    }
}

File: cross-platform/src/core/scriptDoc.ts
/**
 * The `.gws` format, ported from `ScriptDoc.swift`.
 *
 * A header of `@directives` and then a body of verbs. The parser is strict on
 * purpose — an unknown directive or verb is an error rather than a line quietly
 * ignored, because a `.gws` that half-parses is a session that half-plays.
 */
import { SplitMix64 } from "./rng.js";

export type StepKind =
  | "say" | "pause" | "hold" | "media" | "bed" | "beat" | "level" | "surf" | "pan" | "use";

export interface Step {
  kind: StepKind;
  text: string;
  seconds: number;
  args: number[];
  /** For `use`: the mode of the referenced segment, empty for the whole thing. */
  option: string;
}

export const makeStep = (
  kind: StepKind, p: Partial<Omit<Step, "kind">> = {},
): Step => ({ kind, text: p.text ?? "", seconds: p.seconds ?? 0, args: p.args ?? [], option: p.option ?? "" });

export interface ScriptDoc {
  title: string;
  level: string;
  voice: string;
  /** `return` or `stay`. */
  ending: string;
  seed?: bigint | undefined;
  pan: number;
  beatOverride?: number;
  carrierOverride?: number;
  segment?: string;
  /** Structural density, 1–3. Distinct from variants, which are the same
   *  structure phrased differently. */
  verbosity?: number;
  levels: string[];
  provisional: boolean;
  family?: string;
  from?: string;
  duration: string;
  protectedTerms: string[];
  fixed: boolean;
  continuousExit: boolean;
  continuousExitDefault: boolean;
  shelved?: string;
  upright: boolean;
  needs: string[];
  steps: Step[];
}

export function emptyDoc(): ScriptDoc {
  return {
    title: "untitled", level: "F10", voice: "default", ending: "return",
    pan: 0, levels: [], provisional: false, duration: "", protectedTerms: [],
    fixed: false, continuousExit: false, continuousExitDefault: false,
    upright: false, needs: [], steps: [],
  };
}

export class ScriptError extends Error {
  constructor(readonly kind: string, message: string) { super(message); this.name = "ScriptError"; }
  static directiveAfterBody = (l: string) => new ScriptError("directiveAfterBody", `directive after the body started: ${l}`);
  static unknownDirective = (k: string) => new ScriptError("unknownDirective", `unknown directive @${k}`);
  static unknownVerb = (v: string, line: string) => new ScriptError("unknownVerb", `unknown verb '${v}' in line: ${line}`);
  static badEnding = (v: string) => new ScriptError("badEnding", `@ending must be 'return' or 'stay', got '${v}'`);
  static badNumber = (s: string) => new ScriptError("badNumber", `expected a number, got '${s}'`);
  static badVerbosity = (s: string) => new ScriptError("badVerbosity", `@verbosity must be 1, 2 or 3, got '${s}'`);
  static variantsInFixed = (s: string) => new ScriptError("variantsInFixed", `@fixed script contains a variant group: ${s}`);
}

const panWords: Record<string, number> = { left: -0.9, right: 0.9, centre: 0, center: 0 };

export function parsePan(v: string): number {
  const s = v.trim().toLowerCase();
  const w = panWords[s];
  if (w !== undefined) return w;
  const d = swiftDouble(s);
  if (d === undefined) throw ScriptError.badNumber(s);
  return d;
}

/**
 * `Double(_ text: String)` semantics, which are stricter than `parseFloat`.
 *
 * Swift rejects trailing rubbish outright — `Double("3s")` is nil — where
 * `parseFloat("3s")` happily returns 3. A `.gws` line reading `pause 3s` must
 * be an error on both platforms or the two disagree about what is a valid file.
 */
export function swiftDouble(s: string): number | undefined {
  const t = s.trim();
  if (t === "") return undefined;
  if (!/^[+-]?((\d+\.?\d*)|(\.\d+))([eE][+-]?\d+)?$/.test(t)
      && !/^[+-]?0[xX][0-9a-fA-F]+(\.[0-9a-fA-F]*)?([pP][+-]?\d+)?$/.test(t)) {
    return undefined;
  }
  const d = Number(t);
  return Number.isNaN(d) ? undefined : d;
}

function num(s: string): number {
  const d = swiftDouble(s.trim());
  if (d === undefined) throw ScriptError.badNumber(s);
  return d;
}

/** `split(separator:maxSplits:)` on a single space, Swift-style: leading and
 *  repeated separators produce empty pieces that Swift then drops. */
function splitFirst(line: string): [string, string] {
  const parts = line.split(" ").filter(p => p !== "");
  if (parts.length === 0) return ["", ""];
  const head = parts[0]!;
  const rest = line.slice(line.indexOf(head) + head.length).trim();
  return [head, rest];
}

export function parse(source: string, seedOverride?: bigint): ScriptDoc {
  const doc = emptyDoc();
  let headerDone = false;

  for (const raw of source.split("\n")) {
    const line = raw.trim();
    if (line === "" || line.startsWith("#")) continue;

    if (line.startsWith("@")) {
      if (headerDone) throw ScriptError.directiveAfterBody(line);
      const [rawKey, val] = splitFirst(line.slice(1));
      const key = rawKey.toLowerCase();
      switch (key) {
        case "title": doc.title = val; break;
        case "level": doc.level = val; break;
        case "voice": doc.voice = val; break;
        case "ending":
          if (val !== "return" && val !== "stay") throw ScriptError.badEnding(val);
          doc.ending = val; break;
        // `UInt64(val)` is nil for anything that is not a plain unsigned
        // integer, and a nil seed is not an error — it means unseeded.
        case "seed": doc.seed = /^\d+$/.test(val) ? BigInt(val) : undefined; break;
        case "beat": doc.beatOverride = num(val); break;
        case "carrier": doc.carrierOverride = num(val); break;
        case "pan": doc.pan = parsePan(val); break;
        case "segment": doc.segment = val; break;
        case "verbosity": {
          const v = /^\d+$/.test(val) ? Number(val) : NaN;
          if (!(v >= 1 && v <= 3)) throw ScriptError.badVerbosity(val);
          doc.verbosity = v; break;
        }
        case "levels": doc.levels = val.split(",").map(s => s.trim()); break;
        case "from": doc.from = val.toUpperCase(); break;
        case "family": doc.family = val; break;
        case "provisional": doc.provisional = true; break;
        case "duration": doc.duration = val; break;
        case "upright": doc.upright = true; break;
        case "needs": doc.needs = val.split(",").map(s => s.trim()).filter(s => s !== ""); break;
        case "protected": doc.protectedTerms = val.split(",").map(s => s.trim()); break;
        case "fixed": doc.fixed = true; break;
        case "continuous-exit":
          doc.continuousExit = true;
          switch (val.toLowerCase()) {
            case "": break;
            case "default": doc.continuousExitDefault = true; break;
            default: throw ScriptError.unknownDirective(`continuous-exit ${val}`);
          }
          break;
        case "shelved": doc.shelved = val === "" ? "intentionally inactive" : val; break;
        default: throw ScriptError.unknownDirective(key);
      }
      continue;
    }

    headerDone = true;
    const [rawVerb, rest] = splitFirst(line);
    const verb = rawVerb.toLowerCase();
    switch (verb) {
      case "say": doc.steps.push(makeStep("say", { text: rest })); break;
      case "pause": doc.steps.push(makeStep("pause", { seconds: num(rest) })); break;
      case "hold": doc.steps.push(makeStep("hold", { seconds: num(rest) })); break;
      case "media": {
        const f = rest.split(" ").filter(s => s !== "");
        if (f.length !== 2) throw ScriptError.badNumber(rest);
        doc.steps.push(makeStep("media", { text: f[0]!, seconds: num(f[1]!) }));
        break;
      }
      case "level": doc.steps.push(makeStep("level", { text: rest })); break;
      case "beat": doc.steps.push(makeStep("beat", { args: [num(rest)] })); break;
      case "surf": doc.steps.push(makeStep("surf", { args: [num(rest)] })); break;
      case "pan": doc.steps.push(makeStep("pan", { args: [parsePan(rest)] })); break;
      case "use": {
        const [id, option] = splitFirst(rest);
        doc.steps.push(makeStep("use", { text: id, option }));
        break;
      }
      case "bed": {
        const f = rest.split(" ").filter(s => s !== "");
        if (f.length < 2) throw ScriptError.badNumber(rest);
        doc.steps.push(makeStep("bed", { args: [num(f[0]!), num(f[1]!)] }));
        break;
      }
      default: throw ScriptError.unknownVerb(verb, line);
    }
  }

  const seed = seedOverride ?? doc.seed;
  const rng = new SplitMix64(seed ?? 0n);
  for (const step of doc.steps) {
    if (step.kind !== "say") continue;
    const t = step.text;
    if (doc.fixed && t.includes("{")) throw ScriptError.variantsInFixed(t);
    step.text = doc.fixed ? tidy(t) : resolveVariants(t, rng);
  }
  doc.seed = seed;
  return doc;
}

/** Innermost group = the last `{` with no `{` after it before its `}`. */
function lastIndexOfOpen(s: string): number | undefined {
  let found: number | undefined;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "{") found = i;
    if (s[i] === "}" && found !== undefined) return found;
  }
  return undefined;
}

/** Collapse `{a|b|c}` groups, innermost first. Six passes, as in Swift. */
export function resolveVariants(text: string, rng: SplitMix64): string {
  let out = text;
  for (let n = 0; n < 6; n++) {
    const open = lastIndexOfOpen(out);
    if (open === undefined) break;
    const closeRel = out.indexOf("}", open);
    if (closeRel < 0) break;
    const inner = out.slice(open + 1, closeRel);
    const opts = inner.split("|").map(s => s.trim());
    const pick = opts.length === 0 ? "" : opts[Number(rng.next() % BigInt(opts.length))]!;
    out = out.slice(0, open) + pick + out.slice(closeRel + 1);
  }
  return tidy(out);
}

export function fill(text: string, values: Record<string, string>): string {
  let out = text;
  for (const [k, v] of Object.entries(values)) out = out.split(`[[${k}]]`).join(v);
  return tidy(out);
}

export function tidy(s: string): string {
  let out = s;
  while (out.includes("  ")) out = out.split("  ").join(" ");
  return out.trim();
}

/** Tokens still present in spoken text. A line still carrying `[[destination]]`
 *  would be read out loud, so this must be empty before rendering. */
export function unfilledTokens(doc: ScriptDoc): string[] {
  const found: string[] = [];
  for (const step of doc.steps) {
    if (step.kind !== "say") continue;
    let rest = step.text;
    for (;;) {
      const open = rest.indexOf("[[");
      if (open < 0) break;
      const close = rest.indexOf("]]", open + 2);
      if (close < 0) break;
      found.push(rest.slice(open + 2, close));
      rest = rest.slice(close + 2);
    }
  }
  return found;
}

export function filled(doc: ScriptDoc, values: Record<string, string>): ScriptDoc {
  return { ...doc, steps: doc.steps.map(s => s.kind === "say" ? { ...s, text: fill(s.text, values) } : s) };
}

/** Terms that must appear verbatim. Returns the ones that went missing. */
export function missingProtectedTerms(doc: ScriptDoc): string[] {
  if (doc.protectedTerms.length === 0) return [];
  const body = doc.steps.filter(s => s.kind === "say").map(s => s.text).join(" ").toLowerCase();
  return doc.protectedTerms.filter(t => !body.includes(t.toLowerCase()));
}
